docs/Technical-Documentation/tech-memos/priotitized-errors/prioritized-errors.md

```diff
@@ -0,0 +1,101 @@
+
+```python
+paginator = Paginator(parser_errors, settings.BULK_CREATE_BATCH_SIZE)
+row_idx = 6
+for page in paginator:
+    for record in page.object_list:
+        rpt_month_year = str(getattr(record, 'rpt_month_year', None))
+        fields_json = getattr(record, 'fields_json', {})
+
+        worksheet.write(row_idx, 0, record.case_number)
+        worksheet.write(row_idx, 1, rpt_month_year[:4])
+        worksheet.write(row_idx, 2, calendar.month_name[int(rpt_month_year[4:])] if rpt_month_year[4:] else None)
+        worksheet.write(row_idx, 3, format_error_msg(record.error_message, fields_json))
+        worksheet.write(row_idx, 4, record.item_number)
+        worksheet.write(row_idx, 5, friendly_names(fields_json))
+        worksheet.write(row_idx, 6, internal_names(fields_json))
+        worksheet.write(row_idx, 7, record.row_number)
+        worksheet.write(row_idx, 8, str(ParserErrorCategoryChoices(record.error_type).label))
+```
+
+The three helper functions: `format_error_msg`, `friendly_names`, `internal_names` used to write the appropriate fields can be seen below.
+
+```python
+def format_error_msg(error_msg, fields_json):
+    """Format error message."""
+    for key, value in fields_json['friendly_name'].items():
+        error_msg = error_msg.replace(key, value) if value else error_msg
+    return error_msg
+
+
+def friendly_names(fields_json):
+    """Return comma separated string of friendly names."""
+    return ','.join([i for i in fields_json['friendly_name'].values()])
+
+
+def internal_names(fields_json):
+    """Return comma separated string of internal names."""
+    return ','.join([i for i in fields_json['friendly_name'].keys()])
+```
+
+### Prioritized Errors
+[This OneNote](https://gorafttech.sharepoint.com/:o:/s/TDRSResearchDesign/EnIa1Mn4v7pOskW7BLomXhIBxUMlYLRU_f1C0dxemW7dWw?e=m0rNyI) is invaluable to the implementation of prioritized errors. Prioritizing errors could be a very large and technically challenging feature involving new migrations, validation/validator refactors, etc... However, this can all be avoided by making a key insight for each of the category two and category three validators by way of OFA's requirements for them. For the category two case, the OneNote document generically specifies category two validation surrounding: Family Affiliation, Citizenship and Closure reason. Further discussion with OFA indicated that it is important/a priority for a STT to see all category two errors encompassing these fields in their entirety. That makes prioritizing these category two errors extremely easy because the need to query those fields by specific values and expected values is not required. The queries below provide a complete implementation to query all category two errors encompassing those fields.
+
+```python
+# All cat2 errors associated with FAMILY_AFFILIATION and (CITIZENSHIP_STATUS or CLOSURE_REASON)
```

Comment:
🥇 insight

Reply:
This passage also explains the question you had above wrt querying cat2 validators by value.